### PR TITLE
i95 sweep parameter Invalid values

### PR DESCRIPTION
**Pull request by frederikja163**
Stop testplan execution when SetValue fails. #95

### CODE REVIEW

**Comment by frederikja163** [comment]
@rmadsen-ks im not sure if the current output is what we actually want. Currently the error output is the invocation exception, but i think we instead want to output the inner exception in the error. This is similar to the problem in #606 which was fixed in #652 . 
```
00:00:00.876 : TestPlan    : Starting TestPlan 'tmp' on 06/14/2022 12:53:48, 2 of 2 TestSteps enabled.                                                                                                                                                                                                                                                          00:00:00.893 : Resolver    : Loaded assembly System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a from C:\Windows\Microsoft.Net\assembly\GAC_MSIL\System.Net.Http\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Net.Http.dll                                                                                                              00:00:00.953 : Resolver    : Found match for Keysight.OpenTap.Wpf, Version=9.5.0.0, Culture=neutral, PublicKeyToken=null in C:\Users\fredandr\opentap\bin\Debug\Packages\WPF Controls\Keysight.OpenTap.Wpf.dll                                                                                                                                                  00:00:00.957 : Settings    : No settings file exists for TestComponentSetting. A new instance with default values has been created. [654 us]                                                                                                                                                                                                                    00:00:01.032 : Settings    : No settings file exists for TestPlanRunPackageParameterMonitor. A new instance with default values has been created. [613 us]                                                                                                                                                                                                      00:00:01.033 : Settings    : No settings file exists for TestTestPlanRunMonitor. A new instance with default values has been created. [368 us]                                                                                                                                                                                                                  00:00:01.112 : Settings    : No settings file exists for InstrumentSettings. A new instance with default values has been created. [627 us]                                                                                                                                                                                                                      00:00:01.113 : Settings    : No settings file exists for DutSettings. A new instance with default values has been created. [565 us]                                                                                                                                                                                                                             00:00:01.163 : TestPlan    : "Sweep Time Delay" PrePlanRun completed. [3.19 ms]                                                                                                                                                                                                                                                                                 00:00:01.163 : TestPlan    : PrePlanRun Methods completed [17.9 ms]                                                                                                                                                                                                                                                                                             00:00:01.193 : TestPlan    : "Sweep Time Delay" started.                                                                                                                                                                                                                                                                                                        00:00:01.208 : TestPlan    : Error running "Sweep Time Delay": Exception has been thrown by the target of an invocation..                                                                                                                                                                                                                                       00:00:01.210 : TestPlan    : TargetInvocationException: Exception has been thrown by the target of an invocation.                                                                                                                                                                                                                                               00:00:01.236 : TestPlan    :     at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)                                                                                                                                                                                                              00:00:01.236 : TestPlan    :     at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)                                                                                                                                                                                                               00:00:01.236 : TestPlan    :     at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)                                                                                                                                                                                    00:00:01.236 : TestPlan    :     at System.Reflection.RuntimePropertyInfo.SetValue(Object obj, Object value, Object[] index)                                                                                                                                                                                                                                    00:00:01.236 : TestPlan    :     at OpenTap.MemberData.SetValue(Object owner, Object value) in C:\Users\fredandr\opentap\Engine\Reflection\DotNetTypeData.cs:line 120                                                                                                                                                                                           00:00:01.236 : TestPlan    :     at OpenTap.ParameterMemberData.SetValue(Object owner, Object value) in C:\Users\fredandr\opentap\Engine\DynamicMemberTypeDataProvider.cs:line 141                                                                                                                                                                              00:00:01.236 : TestPlan    :     at OpenTap.Plugins.BasicSteps.SweepParameterStep.Run() in C:\Users\fredandr\opentap\BasicSteps\SweepParameterStep.cs:line 270                                                                                                                                                                                                  00:00:01.236 : TestPlan    :     at OpenTap.TestStepExtensions.DoRun(ITestStep Step, TestPlanRun planRun, TestRun parentRun, IEnumerable`1 attachedParameters) in C:\Users\fredandr\opentap\Engine\TestStep.cs:line 907                                                                                                                                         00:00:01.236 : TestPlan    :   Inner ArgumentException: Delay must be a positive value.                                                                                                                                                                                                                                                                         00:00:01.236 : TestPlan    :     at OpenTap.Plugins.BasicSteps.DelayStep.set_DelaySecs(Double value) in C:\Users\fredandr\opentap\BasicSteps\DelayStep.cs:line 20                                                                                                                                                                                               00:00:01.240 : TestPlan    : Exception caught at:                                                                                                                                                                                                                                                                                                               00:00:01.246 : Resolver    : Loaded assembly System.Diagnostics.StackTrace, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a from C:\Windows\Microsoft.Net\assembly\GAC_MSIL\System.Diagnostics.StackTrace\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Diagnostics.StackTrace.dll                                                                    00:00:01.248 : TestPlan    :     at OpenTap.TestStepRun DoRun(OpenTap.ITestStep, OpenTap.TestPlanRun, OpenTap.TestRun, System.Collections.Generic.IEnumerable`1[OpenTap.ResultParameter]) in C:\Users\fredandr\opentap\Engine\TestStep.cs:line 907                                                                                                              00:00:01.248 : TestPlan    :     at failState execTestPlan(OpenTap.TestPlanRun, System.Collections.Generic.IList`1[OpenTap.ITestStep]) in C:\Users\fredandr\opentap\Engine\TestPlanExecution.cs:line 186                                                                                                                                                        00:00:01.248 : TestPlan    :     at OpenTap.TestPlanRun DoExecute(System.Collections.Generic.IEnumerable`1[OpenTap.IResultListener], System.Collections.Generic.IEnumerable`1[OpenTap.ResultParameter], System.Collections.Generic.HashSet`1[OpenTap.ITestStep]) in C:\Users\fredandr\opentap\Engine\TestPlanExecution.cs:line 614                              00:00:01.248 : TestPlan    :     at OpenTap.TestPlanRun executeInContext(System.Collections.Generic.IEnumerable`1[OpenTap.IResultListener], System.Collections.Generic.IEnumerable`1[OpenTap.ResultParameter], System.Collections.Generic.HashSet`1[OpenTap.ITestStep]) in C:\Users\fredandr\opentap\Engine\TestPlanExecution.cs:line 487                       00:00:01.248 : TestPlan    :     ...                                                                                                                                                                                                                                                                                                                            00:00:01.254 : TestPlan    : "Sweep Time Delay" completed with verdict 'Error'. [57.9 ms]                                                                                                                                                                                                                                                                       00:00:01.255 : Test Step   : Break issued from 'Sweep Time Delay' due to verdict Error. See Break Conditions settings.                                                                                                                                                                                                                                          00:00:01.256 : TestPlan    : Test step runs finished. [92.1 ms]                                                                                                                                                                                                                                                                                                 00:00:01.260 : TestPlan    : "Sweep Time Delay" PostPlanRun completed. [1.95 ms]                                                                                                                                                                                                                                                                                00:00:01.283 : Summary     : ------- Summary of test plan started 06/14/2022 12:53:48 -------                                                                                                                                                                                                                                                                   00:00:01.286 : Summary     :  Sweep Time Delay                                 57.9 ms  Error                                                                                                                                                                                                                                                                   00:00:01.286 : Summary     : ----------------------------------------------------------------                                                                                                                                                                                                                                                                   00:00:01.287 : Summary     : ------- Test plan completed with verdict Error in 230 ms -------                                                                                                                                                                                                                                                                   00:00:01.310 : CLI         : CLI action returned exit code: 50 [568 ms]
 ```

**Comment by frederikja163** [comment]
My comment above has now been adressed after talking with @rmadsen-ks 
Now the output is 
```
00:00:03.888 : CLI         : Loaded test plan from C:\Users\fredandr\opentap\bin\Debug\tmp.TapPlan [3.01 s]
00:00:03.889 : Main        : Test Plan: tmp
00:00:03.892 : CLI         : Executing CLI action: package \ check-updates
00:00:03.893 : Settings    : No settings file exists for ResultSettings. A new instance with default values has been created. [3.60 ms]
00:00:03.897 : Settings    : No settings file exists for PackageManagerSettings. A new instance with default values has been created. [1.66 ms]
00:00:03.901 : TestPlan    : -----------------------------------------------------------------
00:00:03.910 : Resolver    : Loaded assembly System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a from C:\Windows\Microsoft.Net\assembly\GAC_MSIL\System.Net.Http\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Net.Http.dll
00:00:03.931 : Resolver    : Loaded assembly System.Runtime.Serialization, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089 from C:\Windows\Microsoft.Net\assembly\GAC_MSIL\System.Runtime.Serialization\v4.0_4.0.0.0__b77a5c561934e089\System.Runtime.Serialization.dll
00:00:03.934 : TestPlan    : Starting TestPlan 'tmp' on 06/14/2022 13:46:19, 2 of 2 TestSteps enabled.
00:00:03.934 : Settings    : No settings file exists for AuthenticationSettings. A new instance with default values has been created. [438 us]
00:00:04.004 : Resolver    : Found match for Keysight.OpenTap.Wpf, Version=9.5.0.0, Culture=neutral, PublicKeyToken=null in C:\Users\fredandr\opentap\bin\Debug\Packages\WPF Controls\Keysight.OpenTap.Wpf.dll
00:00:04.007 : Settings    : No settings file exists for TestComponentSetting. A new instance with default values has been created. [409 us]
00:00:04.061 : Settings    : No settings file exists for TestPlanRunPackageParameterMonitor. A new instance with default values has been created. [600 us]
00:00:04.062 : Settings    : No settings file exists for TestTestPlanRunMonitor. A new instance with default values has been created. [367 us]
00:00:04.140 : Settings    : No settings file exists for InstrumentSettings. A new instance with default values has been created. [3.12 ms]
00:00:04.141 : Settings    : No settings file exists for DutSettings. A new instance with default values has been created. [844 us]
00:00:04.161 : Resolver    : Loaded assembly System.IO.Compression, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089 from C:\Windows\Microsoft.Net\assembly\GAC_MSIL\System.IO.Compression\v4.0_4.0.0.0__b77a5c561934e089\System.IO.Compression.dll
00:00:04.208 : TestPlan    : "Sweep Time Delay" PrePlanRun completed. [4.36 ms]
00:00:04.208 : TestPlan    : PrePlanRun Methods completed [19.1 ms]
00:00:04.233 : TestPlan    : "Sweep Time Delay" started.
00:00:04.241 : TestPlan    : Error running "Sweep Time Delay": Invocation exception on row 0: Delay must be a positive value..
00:00:04.242 : TestPlan    : ArgumentException: Invocation exception on row 0: Delay must be a positive value.
00:00:04.251 : TestPlan    :     at OpenTap.Plugins.BasicSteps.SweepParameterStep.Run() in C:\Users\fredandr\opentap\BasicSteps\SweepParameterStep.cs:line 281
00:00:04.251 : TestPlan    :     at OpenTap.TestStepExtensions.DoRun(ITestStep Step, TestPlanRun planRun, TestRun parentRun, IEnumerable`1 attachedParameters) in C:\Users\fredandr\opentap\Engine\TestStep.cs:line 907
00:00:04.251 : TestPlan    :   Inner TargetInvocationException: Exception has been thrown by the target of an invocation.
00:00:04.261 : TestPlan    :     at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
00:00:04.261 : TestPlan    :     at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
00:00:04.261 : TestPlan    :     at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
00:00:04.261 : TestPlan    :     at System.Reflection.RuntimePropertyInfo.SetValue(Object obj, Object value, Object[] index)
00:00:04.261 : TestPlan    :     at OpenTap.MemberData.SetValue(Object owner, Object value) in C:\Users\fredandr\opentap\Engine\Reflection\DotNetTypeData.cs:line 120
00:00:04.261 : TestPlan    :     at OpenTap.ParameterMemberData.SetValue(Object owner, Object value) in C:\Users\fredandr\opentap\Engine\DynamicMemberTypeDataProvider.cs:line 141
00:00:04.261 : TestPlan    :     at OpenTap.Plugins.BasicSteps.SweepParameterStep.Run() in C:\Users\fredandr\opentap\BasicSteps\SweepParameterStep.cs:line 273
00:00:04.261 : TestPlan    :   Inner ArgumentException: Delay must be a positive value.
00:00:04.261 : TestPlan    :     at OpenTap.Plugins.BasicSteps.DelayStep.set_DelaySecs(Double value) in C:\Users\fredandr\opentap\BasicSteps\DelayStep.cs:line 20
00:00:04.267 : TestPlan    : Exception caught at:
00:00:04.274 : Resolver    : Loaded assembly System.Diagnostics.StackTrace, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a from C:\Windows\Microsoft.Net\assembly\GAC_MSIL\System.Diagnostics.StackTrace\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Diagnostics.StackTrace.dll
00:00:04.276 : TestPlan    :     at OpenTap.TestStepRun DoRun(OpenTap.ITestStep, OpenTap.TestPlanRun, OpenTap.TestRun, System.Collections.Generic.IEnumerable`1[OpenTap.ResultParameter]) in C:\Users\fredandr\opentap\Engine\TestStep.cs:line 907
00:00:04.276 : TestPlan    :     at failState execTestPlan(OpenTap.TestPlanRun, System.Collections.Generic.IList`1[OpenTap.ITestStep]) in C:\Users\fredandr\opentap\Engine\TestPlanExecution.cs:line 186
00:00:04.276 : TestPlan    :     at OpenTap.TestPlanRun DoExecute(System.Collections.Generic.IEnumerable`1[OpenTap.IResultListener], System.Collections.Generic.IEnumerable`1[OpenTap.ResultParameter], System.Collections.Generic.HashSet`1[OpenTap.ITestStep]) in C:\Users\fredandr\opentap\Engine\TestPlanExecution.cs:line 614
00:00:04.276 : TestPlan    :     at OpenTap.TestPlanRun executeInContext(System.Collections.Generic.IEnumerable`1[OpenTap.IResultListener], System.Collections.Generic.IEnumerable`1[OpenTap.ResultParameter], System.Collections.Generic.HashSet`1[OpenTap.ITestStep]) in C:\Users\fredandr\opentap\Engine\TestPlanExecution.cs:line 487
00:00:04.276 : TestPlan    :     ...
00:00:04.279 : TestPlan    : "Sweep Time Delay" completed with verdict 'Error'. [44.6 ms]
00:00:04.281 : Test Step   : Break issued from 'Sweep Time Delay' due to verdict Error. See Break Conditions settings.
00:00:04.281 : TestPlan    : Test step runs finished. [72.8 ms]
00:00:04.284 : TestPlan    : "Sweep Time Delay" PostPlanRun completed. [1.35 ms]
00:00:04.299 : Summary     : ------- Summary of test plan started 06/14/2022 13:46:19 -------
00:00:04.302 : Summary     :  Sweep Time Delay                                 44.6 ms  Error
00:00:04.303 : Summary     : ----------------------------------------------------------------
00:00:04.304 : Summary     : ------- Test plan completed with verdict Error in 224 ms -------
00:00:04.331 : CLI         : CLI action returned exit code: 50 [3.47 s]
```
I.e. we output the entire stack trace, along with information on which row is failing.

File: BasicSteps/SweepParameterStep.cs [hunk 1]
```diff
@@ -273,6 +274,10 @@ public override void Run()
                     }
                     catch (TargetInvocationException ex)
                     {
+                        if (ex.InnerException is ArgumentException arEx)
```
Comment (rmadsen-ks) [inline BasicSteps/SweepParameterStep.cs:277]: What about this? - its not always ArgumentException.
```suggestion
                        if (ex.InnerException is Exception arEx)
```

**Review by rmadsen-ks — CHANGES_REQUESTED** [review]
Could you come up with a unit test for ensuring this behavior?

**Review by rmadsen-ks — CHANGES_REQUESTED** [review]
Let's also remove the call to UpgradeVerdict that is implied when an exception is thrown.

**Comment by rmadsen-ks** [comment]
It seems there is some code which resets to the initial values:
```
 for (int i = 0; i < sets.Length; i++)
sets[i].SetValue(this, originalValues[i]);
```

maybe that should be in finally clause?

**Pull request merged**